### PR TITLE
ddl: restricting altering the utf8 charset to utf8mb4

### PR DESCRIPTION
**Pull request by ciscoxll**
### What problem does this PR solve? 

Only altering the utf8 charset to utf8mb4.
```
mysql> drop table if exists t;
Query OK, 0 rows affected (0.02 sec)

mysql> create table t(a varchar(10) charset binary);
Query OK, 0 rows affected (0.02 sec)

mysql> alter table t modify column a varchar(10) charset utf8 collate utf8_bin;
ERROR 1105 (HY000): unsupported modify original binary charset modification is not supported
```

```
mysql> create table t(a varchar(10)) charset binary;
Query OK, 0 rows affected (0.01 sec)

mysql> show create table t;
+-------+---------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `a` varchar(10) CHARSET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=binary COLLATE=binary |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> alter table t charset utf8 collate utf8_bin;
ERROR 1105 (HY000): unsupported modify original binary charset modification is not supported, charset from binary to utf8
```
### What is changed and how it works?

1. If the character set is a binary type column, inserting data into it is not verified.
2. So going from binary to other character sets can cause errors.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by ciscoxll** [comment]
@winkyao PTAL.

**Comment by ciscoxll** [comment]
```
mysql> create table t (id int, c int);
Query OK, 0 rows affected (0.00 sec)

mysql> show create table t;
+-------+----------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                           |
+-------+----------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `id` int(11) DEFAULT NULL,
  `c` int(11) DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+----------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
Now the master branch column has no default backfilled character set.

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -1995,6 +1995,14 @@ func (d *ddl) getModifiableColumnJob(ctx sessionctx.Context, ident ast.Ident, or
 	if col == nil {
 		return nil, infoschema.ErrColumnNotExists.GenWithStackByArgs(originalColName, ident.Name)
 	}
+	switch col.Tp {
+	case mysql.TypeString, mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeBlob, mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob, mysql.TypeEnum, mysql.TypeSet:
```
Comment (winkyao) [inline ddl/ddl_api.go]: Why we need to check the col.Tp?

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -2245,6 +2253,11 @@ func (d *ddl) AlterTableCharsetAndCollate(ctx sessionctx.Context, ident ast.Iden
 		// nothing to do.
 		return nil
 	}
+	// Not allow to change binary to utf8mb4 or utf8.
+	if origCharset == charset.CharsetBin {
```
Comment (winkyao) [inline ddl/ddl_api.go]: I think you should put the code into the `modifiableCharsetAndCollation`

File: ddl/ddl_api.go (hunk 1 above)
Comment (ciscoxll) [reply ddl/ddl_api.go]: Needed because the integer type defaults to the binary character set.

File: ddl/ddl_api.go (hunk 2 above)
Comment (ciscoxll) [reply ddl/ddl_api.go]: Put the code into the modifiableCharsetAndCollation need to check the col.Tp.

Comment (winkyao) [reply ddl/ddl_api.go]: You should check the altering charset only if the original charset is not equal to the changing charset, then the changing charset can’t be binary.

File: ddl/db_integration_test.go [hunk 3]
```diff
@@ -527,24 +527,43 @@ func (s *testIntegrationSuite) TestNullGeneratedColumn(c *C) {
...
+	tk.MustExec("use test")
+	tk.MustExec("create table t1(a char(10) charset utf8)")
+	tk.MustExec("alter table t1 modify column a char(10) charset utf8mb4")
+
 	tk.MustExec("create table t(a char(10) charset latin1)")
-	tk.MustExec("alter table t modify column a char(10) charset latin1")
-	tk.MustExec("alter table t modify column a char(10) charset utf8")
-	tk.MustExec("alter table t modify column a char(10) charset utf8mb4")
-	rs, err := tk.Exec("alter table t modify column a char(10) charset utf8mb4 collate utf8bin")
+	rs, err := tk.Exec("alter table t modify column a char(10) charset latin1")
```
Comment (crazycs520) [inline ddl/db_integration_test.go]: Why not use `tk.MustExec`.

File: ddl/db_integration_test.go [hunk 4]
```diff
@@ -527,24 +527,43 @@ func (s *testIntegrationSuite) TestNullGeneratedColumn(c *C) {
...
+	c.Assert(err, IsNil)
+	rs, err = tk.Exec("alter table t modify column a char(10) charset utf8")
 	if rs != nil {
 		rs.Close()
 	}
 	c.Assert(err, NotNil)
-	tk.MustExec("alter table t modify column a char(10) charset utf8mb4 collate utf8mb4_bin")
-	rs, err = tk.Exec("alter table t modify column a char(10) charset utf8 collate utf8_bin")
+
+	rs, err = tk.Exec("alter table t modify column a char(10) charset utf8mb4")
```
Comment (crazycs520) [inline ddl/db_integration_test.go]: How about use `testErrorCode` ?

File: ddl/db_test.go [hunk 5]
```diff
@@ -2063,6 +2063,26 @@ func (s *testDBSuite) TestCheckTooBigFieldLength(c *C) {
...
+func (s *testDBSuite) TestCheckConvertToCharacter(c *C) {
+	s.tk = testkit.NewTestKit(c, s.store)
+	s.tk.MustExec("use test")
+	s.tk.MustExec("drop table if exists t")
+	defer s.tk.MustExec("drop table t")
+	s.tk.MustExec("create table t(a varchar(10) charset binary);")
+	ctx := s.tk.Se.(sessionctx.Context)
+	is := domain.GetDomain(ctx).InfoSchema()
+	t, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	rs, err := s.tk.Exec("alter table t modify column a varchar(10) charset utf8 collate utf8_bin")
```
Comment (crazycs520) [inline ddl/db_test.go]: ditto.

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -1806,10 +1806,8 @@ func modifiableCharsetAndCollation(toCharset, toCollate, origCharset, origCollat
 	if !charset.ValidCharsetAndCollation(toCharset, toCollate) {
 		return ErrUnknownCharacterSet.GenWithStackByArgs(toCharset, toCollate)
 	}
-
-	if toCharset == charset.CharsetUTF8MB4 || (toCharset == charset.CharsetUTF8 && origCharset != charset.CharsetUTF8MB4) {
-		// TiDB treats all the data as utf8mb4, so we support changing the charset to utf8mb4.
-		// And not allow to change utf8mb4 to utf8.
+	if toCharset == charset.CharsetUTF8MB4 && origCharset == charset.CharsetUTF8 {
```
Comment (crazycs520) [inline ddl/ddl_api.go]: This will also restricting alter other charset to utf8/utf8mb4. Such as ascii -> utf8/utf8mb4 is not allow now. Is that right?

Comment (ciscoxll) [reply ddl/ddl_api.go]: Right.

File: ddl/db_integration_test.go (hunk 4 above)
Comment (ciscoxll) [reply ddl/db_integration_test.go]: Oh, `errUnsupportedModifyCharset` testErrorCode is not mysql error code.

File: ddl/db_test.go (hunk 5 above)
Comment (ciscoxll) [reply ddl/db_test.go]: Ditto.

File: ddl/db_integration_test.go [hunk 7]
```diff
@@ -527,24 +527,40 @@ func (s *testIntegrationSuite) TestNullGeneratedColumn(c *C) {
 func (s *testIntegrationSuite) TestChangingCharsetToUtf8(c *C) {
 	tk := testkit.NewTestKit(c, s.store)
 
-	tk.MustExec("USE test")
+	tk.MustExec("use test")
+	tk.MustExec("create table t1(a char(10) charset utf8)")
```
Comment (zimulala) [inline ddl/db_integration_test.go]: Could we insert some records into this table and select the records after changing the charset?

Comment (ciscoxll) [reply ddl/db_integration_test.go]: Okey.

**Pull request merged**